interactive bots: Create Google Calendar bot.

YagoGG opened the pull request:
This implements a bot that allows creating Google Calendar events.

Some notes on the OAuth process and how different calendars would work have been included.

smarx commented:
*Automated message from Dropbox CLA bot*

@YagoGG, it looks like you've already signed the Dropbox CLA. Thanks!

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,233 @@
+            content=msg,
+        ))
+    else:
+        raise 'Unknown message type'
```
This should be `raise Exception('Unknown message type')`.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,233 @@
+                }
+            })
+
+        credentials = get_credentials()
```
May be you would want to do this once in `__init__`.

YagoGG replied:
The thing is that I use `credentials` [here](https://github.com/zulip/zulip/pull/2771/files/8ab6ab503588b3cc80ad03aba2d6d56ec06c7ca0#diff-94006fd9e4c82133827aef81c1463418R81), too. This means I cannot use `__init__`, because `list_calendars()` is outside the class.

What about maybe using a global variable?

umairwaheed commented on contrib_bots/lib/gcalendar.py:
````diff
@@ -0,0 +1,233 @@
+                '{err}\n'
+                '```\n')
+
+            reply_message(error.format(err=e), message, client)
````
You probably don't want to send the exception to the user. Instead use `logging.error` to log exception and send a user friendly error message to Zulip.

YagoGG replied:
The user-friendly error implies parsing JSON, since the exception's message is a JSON string.

I had it already done, it's easy. But is it worth the library import and resources?

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,233 @@
+        #   Expected formats:
+        #       "Meeting with John" 2003-01-14T13:37:00+01:00 2003-01-14T15:37:00+01:00
+        #       Trip 2003-01-14
+        input_re = '(\w+|"[\S ]+") ((\d{4}-(0[1-9]|1[1-2])-(0[1-9]|1[1-9]|2[1-9]|3[0-1])T(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\+|\-)(0[1-9]|1[1-4])\:[03]0 ?){2}|(\d{4}-(0[1-9]|1[1-2])-(0[1-9]|1[1-9]|2[1-9]|3[0-1])$))'
```
Compile regex once in `__init__`.

umairwaheed replied:
I think you can use [dateparser](https://dateparser.readthedocs.io/en/latest/) to handle this.

YagoGG replied:
Do you think it's worth the extra dependence?

A perk would be that it would allow checking that the start date is before the end date, tho.

umairwaheed commented:
@YagoGG, pretty good code! Left some feedback. There seem to be some linter errors in your code as well.

umairwaheed commented:
Yes, global variable works as well.

On Monday, December 19, 2016, Yago González <notifications@github.com>
wrote:

> *@YagoGG* commented on this pull request.
> ------------------------------
>
> In contrib_bots/lib/gcalendar.py
> <https://github.com/zulip/zulip/pull/2771>:
>
> > +                'end': {
> +                    'date': start_str
> +                }
> +            })
> +        else:
> +            end_str = splits[1]
> +            event.update({
> +                'start': {
> +                    'dateTime': start_str
> +                },
> +                'end': {
> +                    'dateTime': end_str
> +                }
> +            })
> +
> +        credentials = get_credentials()
>
> The thing is that I use credentials here
> <https://github.com/zulip/zulip/pull/2771/files/8ab6ab503588b3cc80ad03aba2d6d56ec06c7ca0#diff-94006fd9e4c82133827aef81c1463418R81>,
> too. This means I cannot use __init__, because list_calendars() is
> outside the class.
>
> What about maybe using a global variable?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/2771>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnK459sNIYGipFGHWULL_A2vlf1vqHLks5rJppsgaJpZM4LQNWi>
> .
>


-- 
Sent from mobile.

YagoGG commented:
@umairwaheed Alright, here's an updated commit that fixes what you mentioned.

For the moment I've compiled the regex in `__init__`. Let me know what you think about whether it's worth the extra dependence or not, and I'll replace it with `dateparser`, if required :)

showell commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,234 @@
+# (through a different option) with the desired calendar.
+#
+# Something like:
+#   $ ./run.py lib/gcalendar.py -l
```
See my discussion on Zulip.  The disconnect here is that you don't need `run.py` to run code in this library.  You only need `run.py` to run the bot.

YagoGG commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,234 @@
+
+class GCalendarHandler(object):
+    def __init__(self):
+        regex = ('(\w+|"[\S ]+") ((\d{4}-(0[1-9]|1[0-2])-(0[1-9]|1[0-9]|2[0-9]|3[0-1])'
```
The already compiled regex is the one in a class level. I don't use the uncopiled one anywhere after compiling it, for the moment.

Is that what you mean?

showell commented:
Mentors: @YagoGG and I had some conversations on Zulip about things like using `dateparser`, so we can wait for his next round of changes before adding more comments.

YagoGG commented:
Alright, this update makes the following changes, as talked with @showell:

- The validation regex has been replaced by the [`dateparser`](https://dateparser.readthedocs.io/en/latest/) module, which is more flexible with the user's date input.
- The `list_calendar()` function has been removed. If at any point the possibility of using a different calendar besides the *primary* one, a new PR should be created.
- A small test has been added, to assure the messages are correctly parsed.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,254 @@
+            content=msg,
+        ))
+    else:
+        raise('Unknown message type')
```
This will probably give error. I think you need to use something like `raise Exception('Unknown message type')`. 

Optional but better way would be to create your own exception class:
```
class UnknownMessageTypeError(Exception):
    pass

# ...

raise UnkownMessageTypeError
```

YagoGG replied:
Is it okay to declare the class there, and immediately after that, call it in the next line?

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,254 @@
+
+    except Exception as e:
+        raise(e)
+        logging.error(e)
```
Everything after `raise` will be unreachable. Move the code before `raise`.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,254 @@
+            end_date = dateparser.parse(splits[2], settings={'RETURN_AS_TIMEZONE_AWARE': True})
+
+    except Exception as e:
+        raise(e)
```
Simply saying `raise` works. `raise` should be the last statement in the exception handler.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,254 @@
+        raise(e)
+        logging.error(e)
+
+        return None, None, None
```
Not needed.

YagoGG replied:
I've done that because in `handle_message()` I report the error to the user when the title is `None`.

YagoGG replied:
Nevermind, I just realized what you were talking about. The `raise` isn't required here, no need to crash the program if the user enters a wrong input.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,254 @@
+                         '> **{title}**, from *{startDate}* to *{endDate}*')
+
+            reply_message(reply.format(title=title, startDate=start_str, endDate=end_str), message, client)
+        except errors.HttpError as e:
```
It's better to only include that code in `try` which can raise `HttpError` exception.

YagoGG replied:
```
event = service.events().insert(calendarId='primary', body=event).execute()
```

Can raise an `HttpError` exception.

umairwaheed replied:
Yep, so only that line should be in `try`.

umairwaheed replied:
Can you remove unrelated lines from the `try`. Like this:
```Python
try:
    event = self.service.events().insert(calendarId='primary', body=event).execute()
except errors.HttpError as e:
    logging.exception(e)
    # Other code
```

umairwaheed replied:
What about this one?

YagoGG replied:
I just want the `Other code` to run if there were no exceptions, that's why I left it there. Should I use a `break` in the exception handler instead?

umairwaheed replied:
Currently, `except` will only catch `errors.HttpError` exceptions which can only come from `event = self.service.events().insert(calendarId='primary', body=event).execute()`. Exceptions should be exceptional cases only :). Like a network going down (`HttpError`). Probably, the other place which can raise an exception is `reply_message`. You can create a separate `try` for that.

YagoGG replied:
What I mean is that if an `HttpError` has ocurred, I don't want the user to get the "event created" message (which is what I do in the code after `event = self.service.events().insert(calendarId='primary', body=event).execute()`.

umairwaheed commented:
@YagoGG, looks cool over all. Some feedback left.

YagoGG commented:
Thanks, @umairwaheed. Just fixed the things you mentioned.

If you have any doubts on what I did, I answered the comments you posted before. Let me know if something isn't clear :)

femnad commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,261 @@
+            content=msg,
+        ))
+    else:
+        class UnknownMessageTypeError(Exception):
```
I think it would be better to define this somewhere else.

YagoGG replied:
Would it be okay just before defining the function?

femnad replied:
I wouldn't but a class definition between function definitions, so you can have it before the first function maybe.

femnad commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,261 @@
+    assert start_date.strftime('%Y-%m-%d') == '2014-10-21'
+    assert end_date is None
+
+    msg = '@gcalendar Meeting with John | Martes 14 de enero de 2003 a las 13:37:00 CET | Martes 14 de enero de 2003 a las 15:01:10 CET'
```
This line seems to be too long.

YagoGG commented:
That should do it, @femnad ;)

femnad commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,263 @@
+CLIENT_SECRET_FILE = '.gcal_client_secret.json'
+APPLICATION_NAME = 'Zulip Google Calendar Bot'
+
+credentials = None
```
Why do you need these here?

YagoGG replied:
Sorry, this was here because before there were other functions that used these credentials. Fixed.

femnad commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,263 @@
+            })
+
+        try:
+            # TODO: Choose calendar ID from somwhere
```
typo in `somewhere`

femnad commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,263 @@
+
+    return title, start_date, end_date
+
+credentials = get_credentials()
```
I'm confused as to how these work. You don't seem to access `service` as global variable, yet it all seems to work fine.

YagoGG replied:
Just moved `credentials`, `http` and `service` inside the handler class, so this should limit the scope of these variables to a more adequate one. See my new commit for details.

YagoGG commented:
@femnad New commit.

- Moved `credentials`, `http` and `service` inside the handler class, to limit their scope to a more reasonable one.
- Fixed the typo you mentioned.

Thanks!

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,260 @@
+CLIENT_SECRET_FILE = '.gcal_client_secret.json'
+APPLICATION_NAME = 'Zulip Google Calendar Bot'
+
+class UnknownMessageTypeError(Exception):
```
Pretty cool!

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,260 @@
+
+            reply_message(reply.format(title=title, startDate=start_str, endDate=end_str), message, client)
+        except errors.HttpError as e:
+            logging.error(e)
```
Use `logging.exception`.

umairwaheed commented:
@YagoGG, lgtm other that `try` issue.

YagoGG commented:
Fixed, @umairwaheed.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,260 @@
+    else:
+        raise UnknownMessageTypeError(message_type)
+
+'''Identifies and parses the different parts of a message sent to the bot
```
Is this comment for the `parse_message` function? If yes, move it inside the function as a docstring comment.

umairwaheed commented:
@YagoGG, some feedback left. Almost done!

YagoGG commented:
@umairwaheed Whoops, my bad.

I've also updated `triage_message` to include the new `client` (see PR #2923 for details).

umairwaheed commented:
@YagoGG, this is left https://github.com/zulip/zulip/pull/2771/files#r93894518.

YagoGG commented:
@umairwaheed Sorry, missed that :sweat_smile:

Left a comment.

umairwaheed commented:
Oh. You can return from the exception handler.

On Tuesday, December 27, 2016, Yago González <notifications@github.com>
wrote:

> *@YagoGG* commented on this pull request.
> ------------------------------
>
> In contrib_bots/lib/gcalendar.py
> <https://github.com/zulip/zulip/pull/2771>:
>
> > +            date_format = '%c %Z'
> +
> +            start_str = start_date.strftime(date_format)
> +            end_str = None
> +
> +            if not end_date:
> +                reply = (':calendar: Full day event created!\n'
> +                         '> **{title}**, on *{startDate}*')
> +
> +            else:
> +                end_str = end_date.strftime(date_format)
> +                reply = (':calendar: Event created!\n'
> +                         '> **{title}**, from *{startDate}* to *{endDate}*')
> +
> +            reply_message(reply.format(title=title, startDate=start_str, endDate=end_str), message, client)
> +        except errors.HttpError as e:
>
> What I mean is that if an HttpError has ocurred, I don't want the user to
> get the "event created" message (which is what I do in the code after event
> = self.service.events().insert(calendarId='primary', body=event).execute()
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/2771>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnK4-V9-5O2gw7Yl-XwDuaS16mpM4USks5rMNGPgaJpZM4LQNWi>
> .
>


-- 
Sent from mobile.

YagoGG commented:
@umairwaheed That should do it, sorry for missing the comment!

trueskawka commented on contrib_bots/lib/gcalendar.py:
````diff
@@ -0,0 +1,262 @@
+            error = ('Unknown message format.\n'
+                     'Usage examples:\n\n'
+                     '```\n'
+                     '@gcalendar "Meeting with John" | Tuesday, 14 Jan 2003 13:37:00 | Tuesday, 14 Jan 2003 15:01:10\n'
````
I don't think you need "" in event names.

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,262 @@
+    '''
+    This plugin facilitates creating Google Calendar events.
+
+    Usage:
```
If not docs, you could create a `@gcalendar help` option, that would list examples of usage and info on the date format.

YagoGG commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,262 @@
+    '''Gets valid user credentials from storage.
+
+    If nothing has been stored, or if the stored credentials are invalid,
+    the OAuth2 flow is completed to obtain the new credentials.
```
The "stored credentials" are the ones already saved, once you've passed the OAuth2 flow. If you have a wrong `.json`, there's no way the program can know (or at least that I can think of).

If the file `~/.gcalbot-credentials/gcalbot.json` is manually corrupted, the following happens in the log:

```
INFO:googleapiclient.discovery:URL being requested: POST https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json
INFO:oauth2client.transport:Refreshing due to a 401 (attempt 1/2)
INFO:oauth2client.client:Refreshing access_token
```

After the token is automatically refreshed, it works without any impact for the user.

---

About the revocation token, I've added in my next commit something that should manage this (at least to inform the user of the situation), even though it's rather unlikely having the token revoked while the bot is running.

If it's revoked and then the bot is run, it will just proceed to the OAuth flow, as if it was a "fresh install".

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,367 @@
+        return self.message
+
+class MessageParseError(Exception):
+    def __init__(self, eid):
```
Use `error_id`.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,367 @@
+
+        details = ''
+
+        if eid is 1:
```
`is` is the identity operator. Use `==`.

umairwaheed replied:
Same for other `is` usages in this function.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,367 @@
+        return self.message
+
+# NOTE ABOUT GOOGLE'S OAUTH
+# Right now there are some duplicates of this function in Zulip.
```
Which function? Move this in comment inside that function as a docstring.

YagoGG replied:
It talks about the `get_credentials()` function, that is below.

It isn't strictly part of the functioon's documentation, but more of a note for future developers and users. That's why I didn't use a docstring.

Should I still move it inside one?

YagoGG replied:
I've rephrased it as `some duplicates of the get_credentials() function`, for the moment.

Let me know if this works for you :)

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,367 @@
+    credential_dir = os.path.join(home_dir, '.gcalbot-credentials')
+    if not os.path.exists(credential_dir):
+        os.makedirs(credential_dir)
+    credential_path = os.path.join(credential_dir,
```
No need to wrap if the length of the line is short.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,367 @@
+
+        return is_gcalendar
+
+    def handle_message(self, message, client, state_handler):
```
You want to reduce the size of this function. The easiest thing to do would be to handle help and error conditions in separate functions.

TomaszKolek replied:
But instead of separate functions I'd go with separate methods. Because all of them are bots-related.

umairwaheed replied:
Yeah, class methods would be better.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,367 @@
+                end_date = end_date.replace(tzinfo=get_localzone())
+
+            event.update({
+                'start': {
```
More Pythonic.
```
'start': {'dateTime': start_date.isoformat()},
'end': {'dateTime': end_date.isoformat()},
```

umairwaheed commented:
@YagoGG, looking good now. You need to make the size of `handle_message` smaller but creating smaller specialized functions.

YagoGG commented:
Done, @umairwaheed.

Left a reply in one of your comments, just drop me a line if further changes are needed ;)

TomaszKolek commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,365 @@
+    '''
+    message_type = original_msg['type']
+
+    if message_type == 'private':
```
could we simplify it to:
```
params = {content: msg, type=message_type}
if message_type = private
    params.update({to: original_msg['sender_email']})
elif
   params.update(....)
else:
...

client.send_message(params)
```

YagoGG commented:
Done! Thanks, @TomaszKolek

trueskawka commented:
Hi @YagoGG!

I was wondering if you rethought your implementation after [this discussion](https://chat.zulip.org/#narrow/near/109192/stream/GCI.20help/topic/interactive.20bots.20question) with @showell and @timabbott - specifically the `run_flow` and `argparser`stuff? Steve mentioned having a separate script that you run to create credentials, similar to the [GCalendar integration](https://github.com/zulip/zulip/pull/2733).

What would be the next steps to improve your bot?

YagoGG commented:
Hey, @trueskawka!

The ideal thing would be to use the same script as the GCalendar integration, yes. However, since it's something common to both this bot and the integration, I'd need #2733 merged before making changes here.

After this PR, #2733, and #2951 are merged, I'd be happy to make a unified script, based on the one created by @fazerlicourice7 in his PRs, that allows creating the authentication for the different Google Account scopes, depending on the user needs.

This would allow the user generating the credentials only for the applications they're going to use (such as GMail, Calendar, etc.).

That's why I included the comment about Google's OAuth in my bot's code, mentioning that right now there is some duplication of the `get_credentials()` function.

Does that make sense to you?

timabbott commented:
I just merged #2733; is #2951 also needed?

YagoGG commented:
#2951 uses Google's OAuth, too. With #2733 is enough, although @fazerlicourice7 will have to remove his `google.py` script from his PR.

Please let me know if I should proceed, and also if it should be in a new PR or in this one.

timabbott commented:
I think you can proceed in this PR.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,356 @@
+
+    def handle_message(self, message, client, state_handler):
+        if message['content'] == '@gcalendar help':
+            help_message(message, client)
```
Maybe `send_help_message`? That way it will be clear that this function sends the message.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,356 @@
+    def handle_message(self, message, client, state_handler):
+        if message['content'] == '@gcalendar help':
+            help_message(message, client)
+
```
You can remove this blank line.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,356 @@
+        # timezone is irrelevant in that type of events
+    except IndexError as e:
+        logging.exception(e)
+        return MessageParseError(1), None, None
```
You probably want to `raise MessageParseError(1)`.

YagoGG replied:
No, I want to return it (we don't want to stop the program because of this).

Later on, the proper error message will be sent to the user, through Zulip. The same happens with the rest of the returns of `MessageParseError` instances.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,356 @@
+    except IndexError as e:
+        logging.exception(e)
+        return MessageParseError(1), None, None
+    except MessageParseError as e:
```
You can simplify your code if you remove this method and catch `MessageParseError` in `handle_message`.

YagoGG replied:
Nice idea! Forget my previous reply, then.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,356 @@
+        return e, None, None
+    except UnknownTimeZoneError as e:
+        logging.exception(e)
+        return MessageParseError(3), None, None
```
Simply `raise MessageParseError(3).`

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,356 @@
+
+            return
+
+        title, start_date, end_date = parse_message(message['content'])
```
Here you can add the exception handler:
```Python
try:
    title, start_date, end_date = parse_message(message['content'])
except MessageParseError as e:
    parse_error_message(title, message, client)
     return
```

YagoGG commented:
Done, @umairwaheed. That `raise` idea was a pretty neat one, thank you for the heads up!

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,351 @@
+            error = (':warning: **Error!**\n Something went wrong.\n'
+                     'Please, try again or check the logs if the issue persists.')
+
+            reply_message(error, message, client)
```
Your `reply_message` can raise exception. Make sure you handle that case.

YagoGG replied:
I added the possibility of `reply_message` raising such an exception just in case someone extended the bot in the future with some other features. Right now, there's no possible way that an `UnknownMessageTypeError` is raised (only if Zulip provided an invalid message type, which would be rather worrying).

Should they be handled anyway? I'm just asking, because it would slightly clutter up the code, since `reply_message` is called multiple times along the bot's code.

umairwaheed replied:
Ok, let it be then.

umairwaheed commented:
Cool @YagoGG, looking good. There is one more comment about `reply_message` exception handling. Thanks!

YagoGG commented:
@timabbott @trueskawka I have finished the script we were talking about. I pushed it in a different commit to keep it separate from the bot itself, at least for the moment (we can squash it later if necessary).

Let me know what you think about it ;)

tommyip commented on api/integrations/google/oauth.py:
```diff
@@ -0,0 +1,67 @@
+
+# Uses Google's Client Library
+#   pip install --upgrade google-api-python-client
+from oauth2client import client
```
`from oauth2client import client, tools`

YagoGG replied:
Fixed, thanks ;)

TomaszKolek commented on api/integrations/google/oauth.py:
```diff
@@ -0,0 +1,66 @@
+                        help='Scopes to use in the authentication.')
+
+    flags = argparse.ArgumentParser(parents=[tools.argparser]).parse_args()
+except ImportError:
```
What might cause import error here?

TomaszKolek commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,331 @@
+class UnknownMessageTypeError(Exception):
+    def __init__(self, msgtype):
+        self.type = msgtype
+        self.message = 'Unknown message type "' + msgtype + '"'
```
Could we user `format` here instead of `+`?

TomaszKolek commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,331 @@
+    def triage_message(self, message, client):
+        original_content = message['content']
+
+        is_gcalendar = original_content.startswith('@gcalendar')
```
I think we can just move it to return statement - defining `is_gcalendar` is not needed.

YagoGG commented:
Thanks, @TomaszKolek! Modifications pushed.

trueskawka commented on contrib_bots/lib/gcalendar.py:
````diff
@@ -0,0 +1,329 @@
+           '```\n'
+           '---\n'
+           'Here is some information about how the dates work:\n'
+           '- If an ambiguous date format is used, **the English one will have preference** (`03/01/2016` will '
````
It's the American date format, not English.

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,329 @@
+
+    Timezones can be specified in both numerical (+00:00) or abbreviated
+    format (UTC).
+    If no timezone is specified, the server's will be considered as default.
```
I think it should be `the server's timezone`, even though it's slightly repetitive.

YagoGG replied:
Would something like "...the local one, taken from the server..." sound better, maybe?

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,329 @@
+
+    Right now it only works for the calendar set as "primary" in your account.
+
+    Please, run the script api/integrations/google/oauth.py before using this bot,
```
I think the path should be put in quotation marks.

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,329 @@
+        except IOError:
+            logging.error('Couldn\'t find valid credentials.\n'
+                          'Run the oauth.py script in'
+                          'api/integrations/google/oauth.py first.')
```
Same as line 196.

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,329 @@
+    def usage(self):
+        return '''
+            This plugin will allow users to create events
+            in a Google Calendar. Users should preface
```
Maybe `in their Google Calendar`, since every user will be using their own credentials?

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,329 @@
+            project in the Google Developers Console
+            (https://console.developers.google.com/start/api?id=calendar),
+            download "client secret" as a JSON file, and run
+            api/integrations/google/oauth.py for the Calendar
```
Same as line 196.

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,329 @@
+from oauth2client import client, tools
+from oauth2client.client import HttpAccessTokenRefreshError
+from oauth2client.file import Storage
+from oauth2client.tools import argparser
```
Did you clean up the imports?

umairwaheed commented on api/integrations/google/oauth.py:
```diff
@@ -0,0 +1,63 @@
+        if credentials:
+            http = credentials.authorize(httplib2.Http())
+            old_scopes = list(credentials.retrieve_scopes(http))
+            SCOPES += old_scopes
```
This is not a constant. Uppercase variables represent constants.

umairwaheed commented on api/integrations/google/oauth.py:
```diff
@@ -0,0 +1,63 @@
+from oauth2client import client, tools
+from oauth2client.file import Storage
+
+parent = tools.argparser
```
Need to add a comment to explain why you are using argument parser from `tools` instead of `argparse`.

umairwaheed commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,328 @@
+        if len(splits) >= 3: # Delimited event
+            end_date = dateparser.parse(splits[2].strip(), settings=settings)
+
+        if not start_date or (len(splits) >= 3 and not end_date):  # Unknown date format
```
This should work without brackets as well.

umairwaheed commented:
@YagoGG looks great. Just a couple of comments. Looks good otherwise. I have not tested your bot manually yet.

YagoGG commented:
Done! I've also added some extra documentation in the code, and stripped the message's content, so trailing whitespaces don't cause any problems.

YagoGG commented:
By the way, here are some commands you may want to check during the manual tests.
Hope you find them useful, and save you some time ;)

### Delimited event with timezones
**Message:** `@gcalendar Meeting | 2017/01/05 15:00 | Jan 5 17:00 | EDT`
**Expected result:** Successful event creation.

### Delimited event without timezone
**Message:** `@gcalendar Lunch with Tim | 13:00 | 15:00`
**Expected result:** Successful event creation (with local timezone).

### Delimited event with timezone (using keywords)
**Message:** `@gcalendar Buy groceries | tomorrow 12:00 | tomorrow 12:30 | -03:00`
**Expected result:** Successful event creation.

### Delimited event without timezone (using keywords)
**Message:** `@gcalendar Walk | in 30 minutes | in 2 hours`
**Expected result:** Successful event creation (with local timezone).

### Delimited event with timezone (mixing keywords and dates)
**Message:** `@gcalendar Burrito party | in 2 days | Viernes, 6 de enero | CET`
**Expected result:** Successful event creation.

### Delimited event with keywords and time
**Message:** `@gcalendar Something | in 2 days 15:30 | in a week`
**Expected result:** Successful event creation (with local timezone).

### Delimited event with start date > end date
**Message:** `@gcalendar Travel to the past | today | yesterday`
**Expected result:** Error reply + API's error message in the log.

### Full day event
**Message:** `@gcalendar Hochzeit | Dienstag 10`
**Expected result:** Successful event creation.

### Full day event (using keywords)
**Message:** `@gcalendar Trip to LA | tomorrow`
**Expected result:** Successful event creation.

### Full day event with a time
> Since it's a full day event, times are ignored

**Message:** `@gcalendar Epiphany | Jan 6 10:00`
**Expected result:** Successful event creation, skipping the date.

### Help message
**Message:** `@gcalendar help`
**Expected result:** Reply with help and usage details.

### Unknown parameters
**Messages:**
```
@gcalendar
@gcalendar foo
```
**Expected result:** *Unknown message format* reply.

### Wrong dates
**Messages:**
```
@gcalendar event | 15:30 | foo
@gcalendar foo | bar | foo | EDT
@gcalendar foo | 15:30 | bar
@gcalendar | bar
@gcalendar ||
```
**Expected result:** *Unknown date format* reply.

### Wrong timezone
**Message:** `@gcalendar event | 20:00 | 21:30 | bar`
**Expeted result:** *Unknown timezone* reply.

showell commented:
Closing.  This can be reopened if there is follow-up.

YagoGG commented:
Hi, @showell! I'd like to keep this open (I still want to purse this PR's merging).

What's the next step? I was awaiting more feedback :smile:

timabbott commented:
@trueskawka can you give a round of feedback on this PR when you get a chance?

YagoGG commented:
Of course only if you stll think this is useful for Zulip!
I don't want to force you to merge code if you don't want it :)

timabbott commented:
I think this is a great bot to have in our library.  Over the next few months, we'll want to invest in the infrastructure around bots, and it's definitely useful to have a decent-sized set of bots to work with.

trueskawka commented on api/integrations/google/oauth.py:
```diff
@@ -0,0 +1,80 @@
+
+# This parent argparser is used because it already contains
+# the arguments that Google's Client Library method "tools.run_flow"
+# supports.
```
I think this description also needs information on how to provide the correct `--scopes` argument. It took me a few minutes to find `https://www.googleapis.com/auth/calendar.readonly` in the deleted file from this PR.

trueskawka replied:
Should've expected it from the `.readonly` that using the above scopes gives traceback:
```
INFO:googleapiclient.discovery:URL being requested: POST https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json
WARNING:googleapiclient.http:Encountered 403 Forbidden with reason "insufficientPermissions"
ERROR:root:<HttpError 403 when requesting https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json returned "Insufficient Permission">
Traceback (most recent call last):
  File "/Users/alicja/Google Drive/Zulip/zulip/contrib_bots/lib/gcalendar.py", line 269, in handle_message
    event = self.service.events().insert(calendarId='primary', body=event).execute()
  File "/usr/local/lib/python2.7/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/googleapiclient/http.py", line 838, in execute
    raise HttpError(resp, content, uri=self.uri)
HttpError: <HttpError 403 when requesting https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json returned "Insufficient Permission">
INFO:root:waiting for next message
```

Wondering how it was working earlier.

trueskawka replied:
Worked after I removed `.readonly` from the argument.

YagoGG replied:
No information on specific scopes is given here, because this script is intented for usage with any other integration that uses Google services. Thus, each bot/integration will need different scopes, depending on its needs. That also explains why it's in `api/integrations/google/`, and not with my bot's code (this was initially created by @fazerlicourice7).

If you take a look at the bot's code itself (`contrib_bots/gcalendar.py`), you'll see that the correct scope is indicated in the setup comments:

```
# Modify this to match where your credentials are stored.
# It should be the same path you used in the
# '--credential_path' argument, when running the script
# in "api/integrations/google/oauth.py".
#
# $ python api/integrations/google/oauth.py \
#   --secret_path <your_secret_file_path> \
#   --credential_path <your_credential_path> \
#   -s https://www.googleapis.com/auth/calendar
#
# If you didn't specify any, it should be in the default
# path (~/.google_credentials.json)
CREDENTIAL_PATH = '~/.google_credentials.json'
```

As well as later on, in the bot's `usage()` docstring:

```python
return '''
            This plugin will allow users to create events
            in their Google Calendar. Users should preface
            messages with "@gcalendar".

            Before running this, make sure to register a new
            project in the Google Developers Console
            (https://console.developers.google.com/start/api?id=calendar),
            download the "client secret" as a JSON file, and run
            "api/integrations/google/oauth.py" for the Calendar
            read-write scope
            (https://www.googleapis.com/auth/calendar).
            '''
```

trueskawka replied:
Agreed on calendar, but I still think the `oauth.py` should have some information on what the `scopes` argument is and how to use it - this file is now not self-explanatory.

trueskawka commented on contrib_bots/lib/gcalendar.py:
```diff
@@ -0,0 +1,333 @@
+
+# Uses Google's Client Library
+#   pip install --upgrade google-api-python-client
+from apiclient import discovery, errors
```
I'm getting this traceback:
```
WARNING:googleapiclient.discovery_cache:file_cache is unavailable when using oauth2client >= 4.0.0
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/googleapiclient/discovery_cache/__init__.py", line 41, in autodetect
    from . import file_cache
  File "/usr/local/lib/python2.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 41, in <module>
    'file_cache is unavailable when using oauth2client >= 4.0.0')
ImportError: file_cache is unavailable when using oauth2client >= 4.0.0
INFO:googleapiclient.discovery:URL being requested: GET https://www.googleapis.com/discovery/v1/apis/calendar/v3/rest
```

YagoGG replied:
It seems to be because of [this issue](https://github.com/google/google-api-python-client/issues/299).

I don't know why they provided a workaround and closed the issue, to be honest. Nevertheless, there seems to be [some interest](https://github.com/google/google-api-python-client/issues/325) in fixing it.

The "silencing fix" the project's members suggested doesn't seem to work:

```python
logging.getLogger('googleapicliet.discovery_cache').setLevel(logging.ERROR)
```

But, anyways, it shouldn't be a problem, since we could classify this as "expected behavior".

---

As a side note, one of the comments in the [first issue](https://github.com/google/google-api-python-client/issues/299) I mentioned before says:

> It seems to be the better way to silence this error, if one does not care about the cache, is to simply specify cache_discovery=False when performing discovery.build() -- this avoids the broken code path.

However, I think we shouldn't disable the cache discovery, just in case the library's developers fix this in the future. What do you think?

trueskawka replied:
If it's to be an expected behavior, we should document it - getting unexpected tracebacks in the console when running the bot may be confusing. Are we using `cache_discovery` actively? If not, why would we want not to disable it, if it can be confusing?

trueskawka commented:
Hi @YagoGG!

I tested the bot manually, added some comments to clarify how to run it.

I've also tested your usage examples and encountered some discrepancies:
* `@gcalendar Walk | in 30 minutes | in 2 hours` - this one is interesting, as it created an event at 17:49-19:19 instead, although the timestamp in the `Event created!` message is right

* `Burrito party | in 2 days | Viernes, 6 de enero | CET` traceback:
```
ERROR:root:<HttpError 400 when requesting https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json returned "The specified time range is empty.">
Traceback (most recent call last):
  File "/Users/alicja/Google Drive/Zulip/zulip/contrib_bots/lib/gcalendar.py", line 269, in handle_message
    event = self.service.events().insert(calendarId='primary', body=event).execute()
  File "/usr/local/lib/python2.7/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/googleapiclient/http.py", line 838, in execute
    raise HttpError(resp, content, uri=self.uri)
HttpError: <HttpError 400 when requesting https://www.googleapis.com/calendar/v3/calendars/primary/events?alt=json returned "The specified time range is empty.">
```

* `@gcalendar Something | in 2 days 15:30 | in a week` - this one created an all-day event from Friday till Wednesday, it doesn't start at 15:30, was that expected?

* `@gcalendar Hochzeit | Dienstag 10` - not sure when this event was supposed to be created - I'd expect Thursday at 10, but it created an event on January 10th

Could you check those? Thanks!

YagoGG commented:
Hey @trueskawka! Thank you for your comments. Here are the explanations for the cases you mentioned, I hope you find them clarifying:

- `@gcalendar Walk | in 30 minutes | in 2 hours` it has taken me a while to notice what was happening: it works as it should, because the difference between the timestamp and the event's date **is the timezone**.

    The timestamp shows a different time because it uses the host's timezone to specify it, while Google Calendar uses *your* timezone to display the events. If you take into account the timezone offset, you'll notice that both the timestamp and the event represent the same moment, but with different references (the timezones).

    ![Timestamp](https://cloud.githubusercontent.com/assets/7356565/22350678/05122c80-e415-11e6-9376-48a2c95eeea8.png)

    See the `UTC` in the timestamp? This is how it looks in my calendar (which is set in CET, aka UTC +01:00):

    ![Calendar event](https://cloud.githubusercontent.com/assets/7356565/22350695/1a423104-e415-11e6-9a8c-20b18c078906.png)


- `@gcalendar Burrito party | in 2 days | Viernes, 6 de enero | CET`: sorry about that! I wrote the example list about 3 weeks ago, and the 6th of January passed.

    Try with `@gcalendar Burrito party | in 2 days | Martes, 28 de febrero | CET`, which points to the end of February.

- `@gcalendar Something | in 2 days 15:30 | in a week` creates a "delimited event" (contains start/end times), that lasts for 5 days.

    Notice that since it covers multiple days, Google Calendar will place it in the top of the calendar, as if it was a full day event. In the event's details you'll be able to see that it has the times in it:

    ![Event details](https://cloud.githubusercontent.com/assets/7356565/22348435/eafafe16-e40b-11e6-9ca8-d0066f0e2508.png)

- About `@gcalendar Hochzeit | Dienstag 10`, the `10` isn't the time, **it's the day of the month**.

    It's the same way we would say "it's Tuesday 10th". *(P.S. If you're going to try this again, keep in mind that next Tuesday/Dienstag is 31st. It took me a while to notice ;))*

Let me know if anything doesn't make sense to you, or you disagree with the behavior. :)

YagoGG commented:
By the way, I pushed an updated commit, with a couple minor changes:

- Make (almost all) lines <= 76 characters long.
- Remove a couple of unnecesary imports.

Since in the last review rounds I had to make some modifications in both the OAuth script and the bot, the script's commit now has a mix of changes. I think the best choice would be to squash both commits, but other approaches are welcome.

brainwane commented:
@YagoGG I'm trying out this branch now. When I go to `/integrations/` I see an entry for Google Calendar, but I click on `#google-calendar` and I don't get documentation about setting it up. Since you're putting this in the contrib bots, maybe it shouldn't go on that /integrations page?

timabbott commented:
@derAnfaenger @YagoGG can the two of you collaborate on getting this bot cleaned up and merged?  I'd love to get this open PR cleared :).

YagoGG commented:
@brainwane That's because what you see in /integrations is an independent integration someone else created, that gives you reminders of the events in a Google Calendar.

Meanwhile, this is a bot that allows you to interactively create new events in your calendars. Both share the Google OAuth script in the integrations directory, hence my modifications there :)

---

@timabbott Updated the PR with the latest changes to the bot API and contrib_bots structure.

Thanks @derAnfaenger for helping me catch up with all the progress we've made in the bot platform!

neiljp commented on api/bots/gcalendar/gcalendar.py:
```diff
@@ -0,0 +1,349 @@
+    def __str__(self):
+        return self.message
+
+def reply_message(msg, original_msg, client):
```
This function can now be substituted by calls to bothandler.send_reply(original_msg, msg) ?
That allows the related Exception to be removed.

YagoGG commented:
Hi @neiljp!

Thank you very much for your review. I just replaced the old `client` with `bot_handler`, and switched to  `send_reply`.

As an extra I also improved the exception management system, to avoid the "Something went wrong" error as much as possible, for the sake of having error messages as insightful as possible.

About the tests, in this case it isn't possible to mock HTTP calls that easily, since my interactions with Google's API is through their own library (`google-api-python-client`).

zulipbot-test commented:
Hello @YagoGG, a Zulip maintainer reviewed this pull request over a week ago, but you haven't updated your pull request since. Please take a look at the requested changes and update your pull request accordingly.

Thank you for your valuable contributions to Zulip!

derAnfaenger commented:
This should be closed since it's pulling against the wrong repo. I opened https://github.com/zulip/python-zulip-api/issues/161 in python-zulip-api for migrating this code.
